File: Github.cs
using System;
using System.Net;
using Mono.Options;
using Newtonsoft.Json.Linq;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace Gth {
  public struct UserItem {
    public int      Count;
    public string[] Arr;

    public UserItem(int count) {
      Count = count;
      Arr   = new string[count];
    }
  }

  public class User {
    private const string API_USERS_URL = "https://api.github.com/users/";
    private const string USER_AGENT    = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36";
    private Regex fetch_patt = new Regex( "(repos|gists|followers|following)" );
    private string url;
    private JObject json;

    public string Name;
    public string Bio;
    public string Link;
    public UserItem Repos;
    public UserItem Gists;
    public UserItem Followers;
    public UserItem Following;

    public User(string name) {
      this.url = API_USERS_URL + name;

      var client = new WebClient();
      client.Headers.Add("user-agent", USER_AGENT);

      var res   = client.DownloadString(this.url);
      json = JObject.Parse(res);

      this.Name = json.Value<string>("name");
      this.Bio  = json.Value<string>("bio");
      this.Link = json.Value<string>("html_url");

      this.Repos     = new UserItem(json.Value<int>("public_repos"));
      this.Gists     = new UserItem(json.Value<int>("public_gists"));
      this.Followers = new UserItem(json.Value<int>("followers"));
      this.Following = new UserItem(json.Value<int>("following"));
    }

    public void Fetch(string thing) {
      if (!fetch_patt.Match(thing).Success) {
        Console.WriteLine($"Unsupported endpoint: {thing}");
        return;
      }

      var url    = this.url +  "/" + thing;
      var client = new WebClient();
      client.Headers.Add("user-agent", USER_AGENT);

      var res = client.DownloadString(url);
      var arr = JArray.Parse(res);

      if (thing == "repos") {
        for (int i = 0; i < arr.Count; i++)
          this.Repos.Arr[i] = arr[i].Value<string>("name");
      } else if (thing == "gists") {
        for (int i = 0; i < arr.Count; i++)
          this.Gists.Arr[i] = arr[i].Value<string>("description");
      } else if (thing == "followers") {
        for (int i = 0; i < arr.Count; i++)
          this.Followers.Arr[i] = arr[i].Value<string>("login");
      } else if (thing == "following") {
        for (int i = 0; i < arr.Count; i++)
          this.Following.Arr[i] = arr[i].Value<string>("login");
      }
    }
  }
}

public class App {
  public static bool followers = false;
  public static bool following = false;
  public static bool repos = false;
  public static bool gists = false;

  public static OptionSet opts = new OptionSet {
    { "f|followers", "Shows user followers", f => followers = f != null },
    { "F|following", "Shows user following", F => following = F != null },
    { "r|repos", "Shows user repos", r => repos = r != null },
    { "g|gists", "Shows user gists", g => gists = g != null }
  };

  public static void Main(string[] argv) {
    List<string> args = opts.Parse(argv);

    if (!(args.Count > 0)) {
      Console.WriteLine("No arguments");
      return;
    } else {
      foreach (string name in args) {
        var user = new Gth.User(name);

        Console.WriteLine($"Name: {user.Name}");
        Console.WriteLine($"Bio: {user.Bio}");
        Console.WriteLine($"Link: {user.Link}");

        Console.WriteLine($"Repos: {user.Repos.Count}");
        if (repos) {
          user.Fetch("repos");

          for (var i = 0; i < user.Repos.Arr.Length; i++)
            Console.WriteLine($"| {i + 1}. {user.Repos.Arr[i]}");
        }

        Console.WriteLine($"Gists: {user.Gists.Count}");
        if (gists) {
          user.Fetch("gists");

          for (var i = 0; i < user.Gists.Arr.Length; i++)
            Console.WriteLine($"| {i + 1}. {user.Gists.Arr[i]}");
        }

        Console.WriteLine($"Followers: {user.Followers.Count}");
        if (followers) {
          user.Fetch("followers");

          for (var i = 0; i < user.Followers.Arr.Length; i++)
            Console.WriteLine($"| @{user.Followers.Arr[i]}");
        }

        Console.WriteLine($"Following: {user.Following.Count}");
        if (following) {
          user.Fetch("following");

          for (var i = 0; i < user.Following.Arr.Length; i++)
            Console.WriteLine($"| @{user.Following.Arr[i]}");
        }

        Console.Write("\n");
      }
    }
  }
}

File: Github.py
import http.client as Client
import json        as JSON
import argparse    as ArgParse
import re          as RegExp

USER_AGENT = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36"
API_HOST = "api.github.com"
API_PATH = "/users/"

class GithubRepo(object):
  def __init__(self, name, desc):
    super(GithubRepo, self).__init__()
    self.Name = name
    self.Desc = desc


class GithubUser(object):
  def __init__(self, username):
    super(GithubUser, self).__init__()
    self.url = API_PATH + username

    conn = Client.HTTPSConnection(API_HOST)
    conn.request("GET", self.url, headers={ "User-Agent": USER_AGENT })

    res = conn.getresponse()
    self.json = JSON.loads(res.read())

    self.Name      = self.json["name"]
    self.Bio       = self.json["bio"]
    self.Link      = self.json["html_url"]
    self.Repos     = { "count": self.json["public_repos"], "arr": [] }
    self.Gists     = { "count": self.json["public_gists"], "arr": [] }
    self.Followers = { "count": self.json["followers"], "arr": [] }
    self.Following = { "count": self.json["following"], "arr": [] }

    res.close()
    conn.close()

  def Fetch(self, kind):
    assert (RegExp.match(r'(repos|gists|followers|following)', kind) is not None), f"Unsupported endpoint: {kind}"

    url  = f"{self.url}/{kind}"
    conn = Client.HTTPSConnection(API_HOST)
    conn.request("GET", url, headers={ "User-Agent": USER_AGENT })

    res = conn.getresponse()
    arr = JSON.loads(res.read())

    if kind == "repos":
      for obj in arr:
        n = obj["name"]
        d = obj["description"]
        self.Repos["arr"].append(GithubRepo(n, d))

    elif kind == "gists":
      for obj in arr:
        self.Gists["arr"].append(obj["description"])

    elif kind == "followers":
      for obj in arr:
        self.Followers["arr"].append(obj["login"])

    elif kind == "following":
      for obj in arr:
        self.Following["arr"].append(obj["login"])

parser = ArgParse.ArgumentParser(
  usage="Github.py [OPTIONS] usernames",
  description="Commandline app demostrating the Github REST API",
  epilog="https://github.com/Miqueas/Github-REST-API-Example"
)

parser.add_argument("usernames", nargs='+')
parser.add_argument("-f", "--followers", action="store_true", default=False)
parser.add_argument("-F", "--following", action="store_true", default=False)
parser.add_argument("-r", "--repos",     action="store_true", default=False)
parser.add_argument("-g", "--gists",     action="store_true", default=False)

args = parser.parse_args()

for name in args.usernames:
  user = GithubUser(name)

  print(f"Name: {user.Name}")
  print(f"Bio: {user.Bio}")
  print(f"Link: {user.Link}")

  print(f"Public repos: {user.Repos['count']}")
  if args.repos:
    user.Fetch("repos")

    for i in range(len(user.Repos["arr"])):
      n = user.Repos['arr'][i].Name
      d = user.Repos['arr'][i].Desc
      print(f"| {i + 1}. {n}: {d}")

  print(f"Public gists: {user.Gists['count']}")
  if args.gists:
    user.Fetch("gists")

    for i in range(len(user.Gists["arr"])):
      print(f"| {i + 1}. {user.Gists['arr'][i]}")

  print(f"Followers: {user.Followers['count']}")
  if args.followers:
    user.Fetch("followers")

    for i in range(len(user.Followers["arr"])):
      print(f"| @{user.Followers['arr'][i]}")

  print(f"Following: {user.Following['count']}")
  if args.following:
    user.Fetch("following")

    for i in range(len(user.Following["arr"])):
      print(f"| @{user.Following['arr'][i]}")

  print()